Mathlib/Topology/Order/ScottTopology.lean

```diff
@@ -0,0 +1,364 @@
+
+/-!
+# Scott topology
+
+This file introduces the Scott topology on a preorder.
+
+## Main definitions
+
+- `DirSupInacc` - a set `u` is said to be inaccessible by directed joins if, when the least upper
+  bound of a directed set `d` lies in `u` then `d` has non-empty intersection with `u`.
+- `DirSupClosed` - a set `s` is said to be closed under directed joins if, whenever a directed set
+  `d` has a least upper bound `a` and is a subset of `s` then `a` also lies in `s`.
+- `Topology.scott` - the Scott topology is defined as the join of the topology of upper sets and the
+  Scott-Hausdorff topology (the topological space where a set `u` is open if, when the least upper
+  bound of a directed set `d` lies in `u` then there is a tail of `d` which is a subset of `u`).
+
+## Main statements
+
+- `Topology.IsScott.isUpperSet_of_isOpen`: Scott open sets are upper.
+- `Topology.IsScott.isLowerSet_of_isClosed`: Scott closed sets are lower.
+- `Topology.IsScott.monotone_of_continuous`: Functions continuous wrt the Scott topology are
+  monotone.
+- `Topology.IsScott.scottContinuous_iff_continuous` - a function is Scott continuous (preserves
+  least upper bounds of directed sets) if and only if it is continuous wrt the Scott topology.
+- `Topology.IsScott.instT0Space` - the Scott topology on a partial order is T₀.
+
+## Implementation notes
+
+A type synonym `WithScott` is introduced and for a preorder `α`, `WithScott α` is made an instance
+of `TopologicalSpace` by the `scott` topology.
+
+We define a mixin class `IsScott` for the class of types which are both a preorder and a
+topology and where the topology is the `scott` topology. It is shown that `WithScott α` is an
+instance of `IsScott`.
+
+A class `Scott` is defined in `Topology/OmegaCompletePartialOrder` and made an instance of a
+topological space by defining the open sets to be those which have characteristic functions which
+are monotone and preserve limits of countable chains (`OmegaCompletePartialOrder.Continuous'`).
+A Scott continuous function between `OmegaCompletePartialOrder`s is always
+`OmegaCompletePartialOrder.Continuous'` (`OmegaCompletePartialOrder.ScottContinuous.continuous'`).
+The converse is true in some special cases, but not in general
+([Domain Theory, 2.2.4][abramsky_gabbay_maibaum_1994]).
+
+## References
+
+* [Abramsky and Jung, *Domain Theory*][abramsky_gabbay_maibaum_1994]
+* [Gierz et al, *A Compendium of Continuous Lattices*][GierzEtAl1980]
+* [Karner, *Continuous monoids and semirings*][Karner2004]
+
+## Tags
+
+Scott topology, preorder
+-/
+
+open Set
+
+variable {α β : Type*}
+
+/-! ### Prerequisite order properties -/
+
+section Preorder
+variable [Preorder α] {s t : Set α}
+
+lemma IsUpperSet.upperBounds_subset (hs : IsUpperSet s) : s.Nonempty → upperBounds s ⊆ s :=
+  fun ⟨_a, ha⟩ _b hb ↦ hs (hb ha) ha
+
+lemma IsLowerSet.lowerBounds_subset (hs : IsLowerSet s) : s.Nonempty → lowerBounds s ⊆ s :=
+  fun ⟨_a, ha⟩ _b hb ↦ hs (hb ha) ha
```

Comment:
These seem fairly general-purpose - should they be moved to the `Preorder` section in `Order/UpperLower/Basic`?